File: go/example_code/ec2/ec2_describe_security_groups.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/awserr"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/ec2"
)

// Describes the security groups by IDs that are passed into the CLI. Takes
// a space separated list of group IDs as input.
//
// Usage:
//
//	go run ec2_describe_security_groups.go groupId1 groupId2 ...
func main() {
	if len(os.Args) < 2 {
		exitErrorf("Security Group ID required\nUsage: %s group_id ...",
			filepath.Base(os.Args[0]))
	}
	groupIds := os.Args[1:]

	// Initialize a session in us-west-2 that the SDK will use to load
	// credentials from the shared credentials file ~/.aws/credentials.
	sess, err := session.NewSession(&aws.Config{
		Region: aws.String("us-west-2")},
	)

	// Create an EC2 service client.
	svc := ec2.New(sess)

	// Retrieve the security group descriptions
	result, err := svc.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
		GroupIds: aws.StringSlice(groupIds),
	})
	if err != nil {
		if aerr, ok := err.(awserr.Error); ok {
			switch aerr.Code() {
			case "InvalidGroupId.Malformed":
				fallthrough
			case "InvalidGroup.NotFound":
				exitErrorf("%s.", aerr.Message())
			}
		}
		exitErrorf("Unable to get descriptions for security groups, %v", err)
	}

	fmt.Println("Security Group:")
	for _, group := range result.SecurityGroups {
		fmt.Println(group)
	}
}

func exitErrorf(msg string, args ...interface{}) {
	fmt.Fprintf(os.Stderr, msg+"\n", args...)
	os.Exit(1)
}

File: javascriptv3/example_code/ec2/src/ec2_describesecuritygroups.ts
/* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0

ABOUT THIS NODE.JS EXAMPLE: This example works with AWS SDK for JavaScript version 3 (v3),
which available at https://github.com/aws/aws-sdk-js-v3. This example is in the 'AWS SDK for JavaScript v3 Developer Guide' at
https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/ec2-example-security-groups.html

Purpose:
ec2_describesecuritygroups.ts demonstrates how to retrieve information about one or more security groups.

Inputs (replace in code):
- REGION
- SECURITY_GROUP_ID

Running the code:
ts-node ec2_describesecuritygroups.ts

 */
// snippet-start:[ec2.JavaScript.SecurityGroups.describeSecurityGroupsV3]

// Import required AWS SDK clients and commands for Node.js
const {
  EC2Client,
  DescribeSecurityGroupsCommand,
} = require("@aws-sdk/client-ec2");

// Set the AWS region
const REGION = "REGION"; //e.g. "us-east-1"

// Create EC2 service object
const ec2client = new EC2Client(REGION);

// Set the parameters
const params = { GroupIds: ["SECURITY_GROUP_ID"] }; //SECURITY_GROUP_ID

const run = async () => {
  try {
    const data = await ec2client.send(
      new DescribeSecurityGroupsCommand(params)
    );
    console.log("Success", JSON.stringify(data.SecurityGroups));
  } catch (err) {
    console.log("Error", err);
  }
};
run();

// snippet-end:[ec2.JavaScript.SecurityGroups.describeSecurityGroupsV3]

